exposed-dao/src/main/kotlin/org/jetbrains/exposed/dao/EntityCache.kt

```diff
@@ -42,29 +46,31 @@ class EntityCache(private val transaction: Transaction) {
     }
 
     fun <ID:Comparable<ID>, T: Entity<ID>> scheduleInsert(f: EntityClass<ID, T>, o: T) {
-        inserts.getOrPut(f.table) { arrayListOf() }.add(o as Entity<*>)
+        inserts.getOrPut(f.table) { LinkedIdentityHashSet() }.add(o as Entity<*>)
+    }
+
+    fun <ID:Comparable<ID>, T: Entity<ID>> scheduleUpdate(f: EntityClass<ID, T>, o: T) {
+        updates.getOrPut(f.table) { LinkedIdentityHashSet() }.add(o as Entity<*>)
     }
 
     fun flush() {
         flush(inserts.keys + data.keys)
     }
 
     private fun updateEntities(idTable: IdTable<*>) {
-        data[idTable]?.let { map ->
-            if (map.isNotEmpty()) {
-                val updatedEntities = HashSet<Entity<*>>()
-                val batch = EntityBatchUpdate(map.values.first().klass)
-                for ((_, entity) in map) {
-                    if (entity.flush(batch)) {
-                        check(entity.klass !is ImmutableEntityClass<*, *>) { "Update on immutable entity ${entity.javaClass.simpleName} ${entity.id}" }
-                        updatedEntities.add(entity)
-                    }
-                }
-                batch.execute(transaction)
-                updatedEntities.forEach {
-                    transaction.registerChange(it.klass, it.id, EntityChangeType.Updated)
+        updates.remove(idTable)?.takeIf { it.isNotEmpty() }?.let {
```

Comment:
This mirrors the `flushInserts` logic for updates.  Examining `data` would check every entity in the cache to see if it needs an update.  A lot of time is saved by annotating which entities need to be updated.